crates/ruff_linter/src/rules/ruff/rules/unused_variable_accessed.rs

```diff
@@ -0,0 +1,131 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, ViolationMetadata};
+use ruff_python_ast::helpers::is_dunder;
+use ruff_python_semantic::{Binding, BindingFlags, BindingKind};
+use ruff_python_stdlib::builtins::is_python_builtin;
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for usages of variables marked as unused (variable names starting with an underscore, except '_') in functions.
+///
+/// ## Why is this bad?
+/// Marking variables with a leading underscore conveys that they are intentionally unused within the function or method.
+/// When these variables are later referenced in the code, it causes confusion and potential misunderstandings about
+/// the code's intention. A variable marked as "unused" but subsequently used suggests oversight or unintentional use
+/// and detracts from the clarity and maintainability of the codebase.
+///
+/// ## Example
+/// ```python
+/// def function():
+///     _variable = 3
+///     return _variable + 1
+/// ```
+///
+/// Use instead:
+/// ```python
+/// def function():
+///     variable = 3
+///     return variable + 1
+/// ```
+#[derive(ViolationMetadata)]
+pub(crate) struct UnusedVariableAccessed {
+    name: String,
+    shadowed_kind: ShadowedKind,
+}
+
+impl Violation for UnusedVariableAccessed {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!(
+            "Local variable `{}` with leading underscore is accessed",
+            self.name
+        )
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some(match self.shadowed_kind {
+            ShadowedKind::BuiltIn => {
+                "Prefer using trailing underscores to avoid shadowing a built-in".to_string()
+            }
+            ShadowedKind::Some => {
+                "Prefer using trailing underscores to avoid shadowing a variable".to_string()
+            }
+            ShadowedKind::None => "Remove leading underscores".to_string(),
+        })
+    }
+}
+
+/// RUF052
+pub(crate) fn unused_variable_accessed(
+    checker: &Checker,
+    binding: &Binding,
+) -> Option<Vec<Diagnostic>> {
+    let name = binding.name(checker.source());
+
+    // only variables marked as private
+    if !name.starts_with('_') || name == "_" || is_dunder(name) {
```

Comment:
I think we actually want this rule to flag all names matching the `dummy_variable_rgx` _even if_ they don't start with a a leading underscore. You take care of the dummy-variable regex later on in this function, so maybe we should just get rid of this first condition here:

```suggestion
    // Ignore `_` and dunder variables
    if name == "_" || is_dunder(name) {
```

Reply:
Should the rule be renamed to dummy-variable-accessed?
If so, then the error messages will need a change.

Reply:
I think that renaming makes sense. The docs probably also need to talk a little bit more about how the `dummy-variable-rgx` can be used to customise the behaviour, as well -- it's fine to focus on the default behaviour, but a sentence or two saying how it can be customised would probably be useful.